core/flow/traffic_shaping_test.go

```diff
@@ -0,0 +1,142 @@
+		rule           *FlowRule
+	}
+
+	rule := []byte(`{"resource":"Sentinel","limitApp":"","grade":1,"count":0,"strategy":0,"controlBehavior":0,"warmUpPeriodSec":10,"maxQueueingTimeMs":1000,"clusterMode":true,"clusterConfig":{"thresholdType":1}}`)
+	flowRule := &FlowRule{}
+	_ = json.Unmarshal(rule, &flowRule)
+
+	tests := []struct {
+		name   string
+		fields fields
+		want   *FlowRule
+	}{
+		{"test1", fields{nil, nil, flowRule}, flowRule},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			traffic := &TrafficShapingController{
+				flowCalculator: tt.fields.flowCalculator,
+				flowChecker:    tt.fields.flowChecker,
+				rule:           tt.fields.rule,
+			}
+			if got := traffic.Rule(); !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("TrafficShapingController.Rule() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestTrafficShapingController_FlowChecker(t *testing.T) {
+	type fields struct {
+		flowCalculator TrafficShapingCalculator
+		flowChecker    TrafficShapingChecker
+		rule           *FlowRule
+	}
+
+	flowChecker := NewThrottlingChecker(1000)
+
+	tests := []struct {
+		name   string
+		fields fields
+		want   TrafficShapingChecker
+	}{
+		{"test1", fields{nil, flowChecker, nil}, flowChecker},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			traffic := &TrafficShapingController{
+				flowCalculator: tt.fields.flowCalculator,
+				flowChecker:    tt.fields.flowChecker,
+				rule:           tt.fields.rule,
+			}
+			if got := traffic.FlowChecker(); !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("TrafficShapingController.FlowChecker() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestTrafficShapingController_PerformChecking(t *testing.T) {
+	type fields struct {
+		flowCalculator TrafficShapingCalculator
+		flowChecker    TrafficShapingChecker
+		rule           *FlowRule
+	}
+	type args struct {
+		node         base.StatNode
+		acquireCount uint32
+		flag         int32
+	}
+
+	flowCalculator := NewDefaultTrafficShapingCalculator(99)
+	flowChecker := NewThrottlingChecker(1000)
+	tokenResult := base.NewTokenResultPass()
+	node := stat.NewBaseStatNode(1, 1000)
+
+	tests := []struct {
+		name   string
+		fields fields
+		args   args
+		want   *base.TokenResult
+	}{
+		{"test1", fields{flowCalculator, flowChecker, nil}, args{}, tokenResult},
+		{"test2", fields{flowCalculator, flowChecker, nil}, args{node, 1, 1000}, tokenResult},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			traffic := &TrafficShapingController{
+				flowCalculator: tt.fields.flowCalculator,
+				flowChecker:    tt.fields.flowChecker,
+				rule:           tt.fields.rule,
+			}
+			if got := traffic.PerformChecking(tt.args.node, tt.args.acquireCount, tt.args.flag); !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("TrafficShapingController.PerformChecking() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestTrafficShapingController_FlowCalculator(t *testing.T) {
```

Comment:
It is not necessary to follow the test table specification completely for some unit tests with few test point, and from my perspective, it is more important to simplify.

Reply:
For Example:
func TestTrafficShapingController_FlowCalculator(t *testing.T) {
        flowCalculator := NewDefaultTrafficShapingCalculator(99)
	traffic := &TrafficShapingController{
		flowCalculator: flowCalculator,
		flowChecker:    nil,
		rule:           nil,
	}
	if got := traffic.FlowCalculator(); !reflect.DeepEqual(got, flowCalculator) {
		t.Errorf("TrafficShapingController.FlowCalculator() = %v, want %v", got, tt.want)
	}
}

Please let me know if you have better idea.
Thanks.